### PR TITLE
Added username recommendations and quicker verification

### PR DESCRIPTION
**Pull request by tliu22**
Also added recommender.py in ocflib

### CODE REVIEW

File: ocfweb/account/recommender.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from ocflib.account.creation import validate_username
```
Comment (matthew-mcallister) [inline ocfweb/account/recommender.py]: This whole file is unused and duplicates https://github.com/ocf/ocflib/pull/102, though there I suggested that the code be permanently moved here. (Oops, I should have left my comments here instead.)

File: ocfweb/account/register.py [hunk 2]
```diff
@@ -21,7 +24,6 @@
 from ocfweb.component.forms import Form
 from ocfweb.component.forms import wrap_validator
 
-
```
Comment (matthew-mcallister) [inline ocfweb/account/register.py]: This newline is supposed to be there.

File: ocfweb/account/templates/account/register/index.html [hunk 3]
```diff
@@ -118,9 +123,76 @@
...
+                <ul class="list-group list-group-flush" id="recommendations" style="margin: 7px"></ul>
+                <button type="button" class="btn btn-secondary" onclick="recommend()">More</button>
+            </div>
+            <hr class="visible-sm" />
 
             {% include "account/partials/password-requirements-sidebar.html" %}
+
         </div>
     </div>
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
```
Comment (matthew-mcallister) [inline ocfweb/account/templates/account/register/index.html]: We already link to jquery in [`base.html`](https://github.com/ocf/ocfweb/blob/master/ocfweb/templates/base.html#L227) ([and we also redundantly vendor it in for some reason](https://github.com/ocf/ocfweb/blob/master/ocfweb/static/js/jquery-1.11.3.min.js)). If you need a different version, we should try and see if we can upgrade it site-wide instead of linking it twice.

File: ocfweb/account/templates/account/register/index.html [hunk 4]
```diff
@@ -118,9 +123,76 @@
...
+                             '<li class="list-group-item list-group-item-success">'+ data[i] +'</li>'
+                         );
+                     }
+                 }
+             });
+        }
+        $(document).ready(function() {
+            recommend();
+        });
+    </script>
```
Comment (matthew-mcallister) [inline ocfweb/account/templates/account/register/index.html]: Can we try to move most/all of this under [`ocfweb/static/js/pages`](https://github.com/ocf/ocfweb/tree/master/ocfweb/static/js/pages/) instead of making it an inline script? Perhaps we could inline only a stub script that sets the `real_name` variable.

File: ocfweb/account/recommender.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+from random import randint
+
+def recommend(first_name, last_name, n):
+    first_name = first_name.lower()
+    last_name = last_name.lower()
+
+    recommendations = []
+    attempts = 0
+    while len(recommendations) < n and attempts < 20:
+        first_rand = randint(1, len(first_name))
```
Comment (abizer) [inline ocfweb/account/recommender.py]: same comments as Matt from the ocflib PR

File: ocfweb/account/register.py [hunk 6]
```diff
@@ -107,6 +109,23 @@ def request_account(request):
         },
     )
 
+def recommend(request):
+    real_name = request.GET.get('real_name', '')
+    first_name, last_name = real_name.split()
```
Comment (matthew-mcallister) [inline ocfweb/account/register.py]: LDAP entries have middle names, so this is just going to crash.

File: ocfweb/account/register.py [hunk 7]
```diff
@@ -107,6 +109,23 @@ def request_account(request):
...
+    rec_lst = recommender.recommend(first_name, last_name, 3)
+    rec_dict = {}
+    for i in range(len(rec_lst)):
+        rec_dict[str(i)] = rec_lst[i] # convert i to string because JS dicts need string keys
```
Comment (matthew-mcallister) [inline ocfweb/account/register.py]: Does this even need to be a map? You can return just a list in a JSON response.

File: ocfweb/account/register.py [hunk 8]
```diff
@@ -107,6 +109,23 @@ def request_account(request):
...
+    return JsonResponse(rec_dict)
+
+def validate(request):
+    real_name = request.GET.get('real_name', '')
+    username = request.GET.get('username', '')
+    try:
+        validate_username(username, real_name)
+        return JsonResponse({'is_valid': True})
+    except:
+        return JsonResponse({'is_valid': False})
```
Comment (matthew-mcallister) [inline ocfweb/account/register.py]: Instead of returning a boolean, can we return a string, i.e. one of `"available"`, `"taken"`, or `"invalid"`? That way we can give more specific feedback on the form.

File: ocfweb/account/register.py (hunk 7 above)
Comment (abizer) [inline ocfweb/account/register.py]: I don't think you need to do this conversion manually, the JsonResponse function should automatically convert this dict to valid JSON. Also, these variables should probably be named properly

File: ocfweb/account/templates/account/register/index.html [hunk 9]
```diff
@@ -118,9 +123,76 @@
...
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
+    <script>
+        $(document).ready(function() {
+            var finTypingCountdown = 250; // 250 milliseconds
+            var typingTimer;
+            var $input = $('#id_ocf_login_name');
+
+            // On keyup, start countdown
+            $input.keyup(function() {
+                //alert('key UP');
```
Comment (abizer) [inline ocfweb/account/templates/account/register/index.html]: remove these debug alerts

**Comment by abizer** [comment]
You probably also want to delete those extraneous vim crud files

File: ocfweb/account/register.py (hunk 7 above)
Comment (tliu22) [reply ocfweb/account/register.py]: screenshot-docs djangoproject com-2017-11-17-09-10-42-640

I wasn't all too sure whether it was okay to make this assumption.

Comment (tliu22) [reply ocfweb/account/register.py]: As in the keys of the dictionary? I was intending this to essentially mirror the surface functionality of an array or a list. But if the potential security concern I listed above is to be disregarded, then I'll just switch to a list which'll be much cleaner.

File: ocfweb/account/recommender.py (hunk 1 above)
Comment (tliu22) [reply ocfweb/account/recommender.py]: screenshot-supernova ocf berkeley edu-8684-2017-11-17-09-39-37-165
screenshot-supernova ocf berkeley edu-8684-2017-11-17-09-39-12-874
This verify text and green/red highlighting occurs after a specified timeout after user has stopped timing (currently 250ms). 
I originally had the recommendations immediately below the username field (as opposed to the sidebar) but if they were listed vertically, they took up too much space, and I didn't like the way they looked when listed horizontally. But this cosmetic judgment is up to a majority consensus.

File: ocfweb/account/register.py (hunk 7 above)
Comment (matthew-mcallister) [reply ocfweb/account/register.py]: It might be better to return a map with a single entry, i.e. `{"recommendations": [...]}`. This is more flexible in the future event that you want to add additional fields to the output.

Comment (tliu22) [reply ocfweb/account/register.py]: Yea that sounds like a better idea

File: ocfweb/account/register.py [hunk 10]
```diff
@@ -108,6 +113,38 @@ def request_account(request):
...
+
+def validate(request):
+    real_name = request.GET.get('real_name', '')
+    username = request.GET.get('username', '')
+    try:
+        validate_username(username, real_name)
+        return JsonResponse({
+            'is_valid': True,
+            'msg': 'Username is available.',
+        })
```
Comment (matthew-mcallister) [inline ocfweb/account/register.py]: Nice, I like this.

File: ocfweb/account/register.py [hunk 11]
```diff
@@ -108,6 +113,38 @@ def request_account(request):
...
+    except (ValidationError, ValidationWarning) as e:
+        msg = str(e)
+        if msg[-1] != '.':
+            msg += '.'
```
Comment (matthew-mcallister) [inline ocfweb/account/register.py]: Nice.

File: ocfweb/account/recommender.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+    recs = []
+
+    for i in range(len(name_fields)):
+        rec = ''
+        for j in range(len(name_fields)):
+            if i == j:
+                rec += name_fields[j][0]
+            else:
+                rec += name_fields[j]
```
Comment (matthew-mcallister) [inline ocfweb/account/recommender.py]: I feel like we can do better than this by systematically generating names in a clever order. I admittedly haven't thought of a way to iterate through _all_ possible names, but we can get a large number of reasonable names using, say, [`itertools.combinations`](https://docs.python.org/3/library/itertools.html#itertools.combinations).

For example, we could try looping over all combinations of names and taking the first `n` letters of those chosen, where `n = 0, 1, 2, ...`:
```python
# Stop at some low n so we don't loop for a million years
for n in range(3):
    for r in range(1, len(name_fields) + 1)
        for c in itertools.combinations(range(len(name_fields)), r):
            try_names = list(name_fields)
            for i in c:
                try_names[i] = try_names[i][:n]
            <test validity of ''.join(try_names)>
```
We'd probably try the full name first of all, though.

File: ocfweb/account/recommender.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        for name_field in name_fields:
+            rand_index = randint(1, len(name_field))
+            rec += name_field[:rand_index]
+        if rec not in recs:
+            try:
+                validate_username(rec, real_name)
+                recs.append(rec)
+            except Exception:
+                pass
+        attempts += 1
```
[the remark this answers is not in the document]
Comment (tliu22) [reply ocfweb/account/recommender.py]: I don't know if the procedural generator alone will suffice, because in the case of my name, none of the procedural suggestions work ("tonyl", "tliu", and "tonyliu" are all already taken)

[the remark this answers is not in the document]
Comment (matthew-mcallister) [reply ocfweb/account/recommender.py]: Well, the problem there is that the first stage doesn't generate enough names as-is, not that we need two methods of generating names.

File: ocfweb/static/js/pages/account/quick_username_validation.js [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+    function validate_username() {
+        $.ajax({
+             type: 'GET',
+             url: 'validate',
+             data: {'username': $input.val(),
+                    'real_name': $('#real_name').html()},
+             success: function(data) {
+                 if(data.is_valid) {
+                     $input.css('border', '5px solid #3c763d');
```
Comment (chriskuehl) [inline ocfweb/static/js/pages/account/quick_username_validation.js]: i think a slightly nicer approach would be to add css classes like `.valid` and `.invalid` and then use `$input.addClass`

in fact, what if you used the bootstrap validation css? https://getbootstrap.com/docs/3.3/css/#forms-control-validation

this css should be on the page, you'd just need to add the right css class names (i think `has-error` and `has-success`)

File: ocfweb/static/js/pages/account/recommend_username.js [hunk 15]
```diff
@@ -0,0 +1,19 @@
+function recommend() {
+    $.ajax({
+         type: 'GET',
+         url: 'recommend',
+         data: {'real_name': $('#real_name').html()},
```
Comment (chriskuehl) [inline ocfweb/static/js/pages/account/recommend_username.js]: same thing, can we use `.text()`?

File: ocfweb/static/js/pages/account/quick_username_validation.js [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+                     $('#valid_username_notif').css('display', 'block');
+                     $('#valid_username_notif').html(data.msg);
```
Comment (chriskuehl) [inline ocfweb/static/js/pages/account/quick_username_validation.js]: can we use `.text(data.msg)` here to avoid potential XSS vulnerabilities?

File: ocfweb/static/js/pages/account/quick_username_validation.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+                     $('#invalid_username_notif').css('display', 'none');
+
+                 } else {
+                     $input.css('border', '5px solid #a94442');
+                     $('#invalid_username_notif').css('display', 'block');
+                     $('#invalid_username_notif').html(data.msg);
```
Comment (chriskuehl) [inline ocfweb/static/js/pages/account/quick_username_validation.js]: can we use `.text(data.msg)` here to avoid potential XSS vulnerabilities?

File: ocfweb/static/js/pages/account/recommend_username.js [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+         success: function(data) {
+             $('#recommendations').empty();
+             recommendations = data['recommendations'];
+             for (var i in recommendations) {
+                 $('#recommendations').append(
+                     '<li class="list-group-item list-group-item-success">'+ recommendations[i] +'</li>'
```
Comment (chriskuehl) [inline ocfweb/static/js/pages/account/recommend_username.js]: can we construct a jquery object here to avoid XSS vulnerabilities?

e.g. (untested, but i think this is right)

```javascript
$('recommendations').append(
    $('<li>')
        .addClass('list-group-item list-group-item-success')
        .text(recommendations[i])
);
```

also -- what if we made these recommendations clickable (click the recommendation to fill in the username box)?

File: ocfweb/account/register.py (hunk 11 above)
Comment (tliu22) [reply ocfweb/account/register.py]: Modifying the exception messages sounds like a better idea. Made a separate ocflib PR for that.

File: ocfweb/account/templates/account/register/index.html [hunk 19]
```diff
@@ -118,9 +119,19 @@ <h4>OCF &hearts; Python, Linux, Open Source</h4>
                 </p>
             </div>
 
-            <hr />
+            <div class="well well-sm">
+                <strong>Available Account Names</strong>
+                <ul class="list-group list-group-flush" id="recommendations" style="margin: 7px"></ul>
+                <button type="button" class="btn btn-secondary" onclick="recommend()">More</button>
+            </div>
+            <hr class="visible-sm" />
```
[the remark this answers is not in the document]
Comment (tliu22) [reply ocfweb/account/templates/account/register/index.html]: 32960829-421ef38a-cb7b-11e7-8373-13a136975e17
32960830-425f315c-cb7b-11e7-8d25-5815353efd81

File: ocfweb/account/recommender.py (hunk 12 above)
Comment (tliu22) [reply ocfweb/account/recommender.py]: What do you think about keeping the first letter of each name intact (for "Tony Liu", keep "t" and "l")?
And also how bout only removing vowels (though that might greatly limit the total number of possible recommendations)?

**Comment by tliu22** [comment]
I shifted the recommendations over from the sidebar, so it's easier to notice and doesn't end up at bottom of page for mobile. I also removed the "More" button for recommendations and switched to a scrollable list instead (seems more convenient). 
screenshot_2017-12-02_15-26-29

I'm not too familiar with HTML/JS practices on UI element padding, so a lot of my values are fixed. Please let me know if there's a better way of doing them.

And as for the recommendations function, I randomized which recs to show because I figure if a username doesn't like a particular rec, he probably won't like adjacent recs (adjacent with respect to the unvalidated_recs list in recommender.py).

**Comment by abizer** [comment]
definitely should have squashed this first before merging
